### PR TITLE
[Profiler] Add pattern to detect if TF32 is available but not used

### PR DESCRIPTION
**Pull request by davidchencsl**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #81740
* #81733
* #81501
* __->__ #81273
* #81056

Summary: This commit adds recording of a global flag allow_tf32 in all torch ops. Then I add a pattern that detects if a user is setting tf32 flag or not, recommend he/she to do so if is not set.

Test Plan: Added a test in test_profiler.py

Differential Revision: [D37799414](https://our.internmc.facebook.com/intern/diff/D37799414)

### CODE REVIEW

**Comment by davidchencsl** [comment]
@davidchencsl has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D37799414).

File: test/test_profiler.py [hunk 1]
```diff
@@ -1553,6 +1556,22 @@ def case4():
             num_matched.append(len(pattern.matched_events()))
         self.assertEqual(num_matched, [i for i, _ in cases])
 
+    @unittest.skipIf(not torch.cuda.is_available(), "CUDA is required")
+    def test_profiler_fp32_matmul_pattern(self):
+        has_tf32 = 1
```
[the remark this answers is not in the document]
Comment (robieta) [reply test/test_profiler.py:1561]: It should probably also be a property of `FP32MatMulPattern`. Maybe make `skip` a `@property`?

File: torch/profiler/_pattern_matcher.py [hunk 2]
```diff
@@ -169,6 +170,35 @@ def match(self, event: _ProfilerEvent):
...
 
 
+class FP32MatMulPattern(Pattern):
+
+    def __init__(self, prof: profile):
+        super().__init__(prof)
+        self.description = (
+            "You are currently using GPU that supports TF32. "
+            "Please enable TF32 by setting 'torch.backends.cuda.matmul.allow_tf32 = True'")
+        for arch in torch.cuda.get_arch_list():
```
Comment (robieta) [inline torch/profiler/_pattern_matcher.py:180]: See comment above

**Review by robieta — COMMENTED** [review]
Again, couple of minor nits but looks good overall.

**Comment by davidchencsl** [comment]
@davidchencsl has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D37799414).

**Comment by davidchencsl** [comment]
@davidchencsl has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D37799414).

**Comment by davidchencsl** [comment]
@davidchencsl has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D37799414).

**Review by robieta — APPROVED** [review]

**Pull request closed without merging**